### PR TITLE
Fix data channel issues

### LINKED ISSUE #348: Out-of-band data channels wrongly created as in-band ones
**Describe the bug**
Out-of-band data channels are wrongly created as in-band data channels and advertized in-line in the SDP negotiation.

**To Reproduce**
1. Add out-of-band data channel `AddDataChannelAsync(42, "dc", ...)` on both peers
2. Observe error in `peerconnection.cc` : "Failed to create a SCTP data channel because the id is already in use or out of range."

**Expected behavior**
Out-of-band data channel are created as such, and therefore not advertized during SDP exchange.

**Additional context**
See `peer_connection.cpp`:299, `webrtc::DataChannelInit::negotiated` is not assigned, keeps its default value of `false`. Fixed in #200 on `experimental/undock` only.

### LINKED ISSUE #349: In-band data channel test failing
**Describe the bug**
The C# in-band data channel test is failing with what appear to be a race condition. Sometimes the opening side's channel stays in `kConnecting` state forever, making the test fail.
https://dev.azure.com/aipmr/MixedReality-WebRTC-CI/_build/results?buildId=11370&view=results

**To Reproduce**
1. See CI above; run the in-band C# test

**Expected behavior**
Test passes

**Additional context**
Not sure what's happening in details:
- Initial investigation shows some likely race condition with the SCTP handshake, although it doesn't explain to me why that would have an impact on setting the `kOpen` state.
- Data channel is functional for all intents and purposes, but its state is wrong.
- Many users reported similar issues on many different implementations/platforms (Firefox, Chrome, etc.) and therefore there is a lot of noise when searching for info, with unrelated bugs leading to the same behavior.

### PR DESCRIPTION
**Pull request by fibann**
Fixes #348. Fixes #349.

### CODE REVIEW

File: libs/mrwebrtc/include/interop_api.h [hunk 1]
```diff
@@ -312,12 +312,22 @@ struct mrsDataChannelAddedInfo {
   const char* label{nullptr};
 };
 
-/// Callback fired when a data channel is added to the peer connection after
-/// being negotiated with the remote peer.
+/// Callback fired when a data channel is added to the peer connection. This is
```
Comment (djee-ms) [inline libs/mrwebrtc/include/interop_api.h:315]: `fired` => `invoked` (polycheck?)

File: libs/mrwebrtc/include/interop_api.h [hunk 2]
```diff
@@ -312,12 +312,22 @@ struct mrsDataChannelAddedInfo {
...
+/// called for both channels that are created locally and ones that are created
+/// by the remote peer.
+///
+/// Use this callback to call |mrsDataChannelRegisterCallbacks| on new data
+/// channels and to start listening for messages/state changes.
+///
+/// The data channel is initially in the |mrsDataChannelState::kConnecting|
+/// state and will transition to |mrsDataChannelState::kOpen| when it is ready
+/// for use. In order to know when the channel state changes you must call
+/// pass a valid |mrsDataChannelStateCallback| to
```
Comment (djee-ms) [inline libs/mrwebrtc/include/interop_api.h:325]:
```suggestion
/// for use. In order to know when the channel state changes you must
/// pass a valid |mrsDataChannelStateCallback| to
```

File: libs/mrwebrtc/src/data_channel.cpp [hunk 3]
```diff
@@ -74,21 +74,29 @@ bool DataChannel::Send(const void* data, size_t size) noexcept {
   if (data_channel_->buffered_amount() + size > GetMaxBufferingSize()) {
     return false;
   }
+
+  auto state = data_channel_->state();
+  RTC_LOG_F(LS_INFO) << state;
```
Comment (djee-ms) [inline libs/mrwebrtc/src/data_channel.cpp:79]: Remove? Or add contextual message and not just the state itself?

File: libs/mrwebrtc/src/data_channel.cpp [hunk 4]
```diff
@@ -74,21 +74,29 @@ bool DataChannel::Send(const void* data, size_t size) noexcept {
...
+
   rtc::CopyOnWriteBuffer bufferStorage((const char*)data, size);
   webrtc::DataBuffer buffer(bufferStorage, /* binary = */ true);
   return data_channel_->Send(buffer);
 }
 
-void DataChannel::OnStateChange() noexcept {
+void DataChannel::FireOnStateChange() const noexcept {
```
Comment (djee-ms) [inline libs/mrwebrtc/src/data_channel.cpp:86]: `Fire` => `Invoke`

File: libs/mrwebrtc/src/interop/global_factory.cpp [hunk 5]
```diff
@@ -182,6 +182,16 @@ rtc::Thread* GlobalFactory::GetWorkerThread() const noexcept {
 #endif  // defined(WINUWP)
 }
 
+rtc::Thread* GlobalFactory::GetSignalingThread() const noexcept {
+  // This only requires init_mutex_ read lock, which must be acquired to access
+  // the singleton instance.
+#if defined(WINUWP)
+  return impl_->signalingThread.get();
```
Comment (djee-ms) [inline libs/mrwebrtc/src/interop/global_factory.cpp:189]: Comment says "Get the WebRTC signaling thread, or NULL if the library is not initialized." but that probably crashes as `impl_` is NULL no?

File: libs/mrwebrtc/test/data_channel_tests.cpp [hunk 6]
```diff
@@ -224,21 +217,13 @@ TEST_P(DataChannelTests, Send) {
       });
   std::function<void(int32_t, int32_t)> state1_cb(
       [&](int32_t state, int32_t id) {
-        ASSERT_EQ(kId, id);
+        if (expected_id >= 0) {
```
Comment (djee-ms) [inline libs/mrwebrtc/test/data_channel_tests.cpp:220]: Why this changed?

File: libs/mrwebrtc/src/data_channel.cpp (hunk 3 above)
Comment (fibann) [reply libs/mrwebrtc/src/data_channel.cpp:79]: This is a debug leftover, will remove.

File: libs/mrwebrtc/src/interop/global_factory.cpp (hunk 5 above)
Comment (fibann) [reply libs/mrwebrtc/src/interop/global_factory.cpp:189]: I copied the workerThread one 😄 Will fix according to the docs.

File: libs/mrwebrtc/test/data_channel_tests.cpp (hunk 6 above)
Comment (fibann) [reply libs/mrwebrtc/test/data_channel_tests.cpp:220]: I am using this callback to handle both out-of-band and in-band channels, see line 260 and 304. Will add a comment to make it clearer.

File: libs/mrwebrtc/include/interop_api.h (hunk 1 above)
Comment (fibann) [reply libs/mrwebrtc/include/interop_api.h]: Fixed

File: libs/mrwebrtc/include/interop_api.h (hunk 2 above)
Comment (fibann) [reply libs/mrwebrtc/include/interop_api.h]: Fixed

File: libs/mrwebrtc/src/data_channel.cpp (hunk 3 above)
Comment (fibann) [reply libs/mrwebrtc/src/data_channel.cpp]: Fixed

File: libs/mrwebrtc/src/data_channel.cpp (hunk 4 above)
Comment (fibann) [reply libs/mrwebrtc/src/data_channel.cpp]: Fixed

File: libs/mrwebrtc/src/interop/global_factory.cpp (hunk 5 above)
Comment (fibann) [reply libs/mrwebrtc/src/interop/global_factory.cpp]: Fixed

File: libs/mrwebrtc/test/data_channel_tests.cpp (hunk 6 above)
Comment (fibann) [reply libs/mrwebrtc/test/data_channel_tests.cpp:223]: Fixed

**Pull request merged**